pkg/daemon/daemon.go

```diff
@@ -199,6 +202,7 @@ const (
 	caBundleFilePath      = "/etc/kubernetes/kubelet-ca.crt"
 	cloudCABundleFilePath = "/etc/kubernetes/static-pod-resources/configmaps/cloud-config/ca-bundle.pem"
 	userCABundleFilePath  = "/etc/pki/ca-trust/source/anchors/openshift-config-user-ca-bundle.crt"
+	kubeConfigPath        = "/etc/kubernetes/kubeconfig"
```

Comment:
Talking to @rphillips, this [`--bootstrap-kubeconfig` target][1] may only be relevant when the kubelet first launches after a node is provisioned.  After that, the kubelet uses [`--kubeconfig=/var/lib/kubelet/kubeconfig`][2].  I think that means you want the machine-config servers to [continue to drop this data into `/etc/kubernetes/kubeconfig`][3].  But that the MCD should be propagating it through to `/var/lib/kubelet/kubeconfig`, and ignoring (or maybe also updating?) `/etc/kubernetes/kubeconfig`.

[1]: https://github.com/openshift/machine-config-operator/blob/854a866ee6a9dd1c6ce64c08e05eb6a16c99aba2/templates/worker/01-worker-kubelet/_base/units/kubelet.service.yaml#L26C1-L26C1
[2]: https://github.com/openshift/machine-config-operator/blob/854a866ee6a9dd1c6ce64c08e05eb6a16c99aba2/templates/worker/01-worker-kubelet/_base/units/kubelet.service.yaml#L27C9-L27C49
[3]: https://github.com/openshift/machine-config-operator/blob/854a866ee6a9dd1c6ce64c08e05eb6a16c99aba2/pkg/server/server.go#L25

Reply:
+1. Both files should be updated.

Reply:
And as an alternative to managing `/var/lib/kubelet/kubeconfig`, [Ryan][1]:

> We will not want MCO to manage `/var/lib/kubelet/kubeconfig`... This file is managed by the Kubelet.
>
> Removing the `/var/lib/kubelet` directory and restarting Kubelet is a common approach to resetting it's certificates. The Kubelet will need to create a CSR, and the controller-manager needs to sign the CSR for the kubelet to be in a 'running' state. This results in the certs being stored in `/var/lib/kubelet/kubeconfig.`

Although that might bump up against [`newNodeWriter` using][2] [`/var/lib/kubelet/kubeconfig`][3] as its kubeconfig, with the MCO learning to fall back to `/var/lib/kubelet/kubeconfig` or just crash-loop until the kubelet repopulates `/var/lib/kubelet/kubeconfig` or something, if the MCD dies while this dance is happening.

[1]: https://issues.redhat.com/browse/OCPBUGS-25821?focusedId=23760030&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-23760030
[2]: https://github.com/openshift/machine-config-operator/blob/8c4ca51b2f9c22c4dc14092166e0d454ba1401b7/pkg/daemon/writer.go#L79
[3]: https://github.com/openshift/machine-config-operator/blob/8c4ca51b2f9c22c4dc14092166e0d454ba1401b7/pkg/daemon/writer.go#L29

Reply:
And more details on the `/var/lib/kubelet` removal option:

1. Update `/etc/kubernetes/kubeconfig`.
2. Remove `/var/lib/kubelet`.
3. In this window, there's space for the still running kubelet to choke on `/var/lib/kubelet`'s absence.
4. `systemctl restart kubelet`.
5. Incoming kubelet repopulates `/var/lib/kubelet`, which depends on it creating a new CSR.  Some installations might need manual CSR approval here?  I'm not familiar with what approval is automatic vs. manual on things like bring-your-own-node clusters.

If we do `/var/lib/kubelet` removal, it might be safer to:

1. Update `/etc/kubernetes/kubeconfig`.
2. `systemctl stop kubelet`.  The MCD pod will keep coasting along.
3. Remove `/var/lib/kubelet`.
4. Because the MCD loads its kubeconfig from ` /var/lib/kubelet/kubeconfig` at the moment, it might need changes to suvive a container restart during this phase.
5. `systemctl start kubelet`.
6. Incoming kubelet repopulates `/var/lib/kubelet`, which depends on it creating a new CSR.  Same CSR approval issues as the other `/var/lib/kubelet` option.

Reply:
Ok, that second approach seems pretty sane to me. I do worry about incurring extra ctr restarts. However, if we need to somehow manage both files, it is better to incur a restart than to delete /var/lib/kubelet and potentially hurt nodeWriter.

so the quick summary here is 

1. manage both /etc/kubernetes/kubeconfig and /var/lib/kubelet/kubeconfig. Write to the /etc file and delete the /var/lib dir so that the kubelet re-populates it. 
2. in order to delete the dir safely, stop and then restart the kubelet. 
3. this should all happen in the daemon's certificate updating loop.
4. the MCS should keep updating just /etc/kubernetes/kubeconfig. 

One question I have on 4 @wking is that in the non-backported version I update the MCS served ignition with the NEW contents of /etc/kubernetes/kubeconfig in the cluster-server (non-bootstrap). Should I do that here too?

Reply:
Is it necessary to remove the whole /var/lib/kubelet directory? When I try to reproduce this step manually I get errors like this one (stop kubelet, rm -rf /var/lib/kubelet as root):

```
[root@sregidor-rot1-9f2sw-worker-a-nnjlw core]# rm -rf /var/lib/kubelet
rm: cannot remove '/var/lib/kubelet/pods/0812ed68-510c-4b31-a618-022b299ade5b/volumes/kubernetes.io~secret/sdn-metrics-certs': Device or resource busy
rm: cannot remove '/var/lib/kubelet/pods/0812ed68-510c-4b31-a618-022b299ade5b/volumes/kubernetes.io~projected/kube-api-access-rvtgk': Device or resource busy

```

And after starting again the kubelet service I get these error messages


```
Mon 2024-01-15 14:08:40 UTC sregidor-rot1-9f2sw-worker-a-nnjlw.c.openshift-qe.internal kubelet.service[157677]: I0115 14:08:40.173287  157677 csi_plugin.go:913] Failed to contact API server when waiting for CSINde publishing: csinodes.storage.k8s.io "sregidor-rot1-9f2sw-worker-a-nnjlw.c.openshift-qe.internal" is forbidden: User "system:anonymous" cannot get resource "csinodes" in API group "storage.k8s.io" at the clustr scope
```

Isn't it enough to remove the /var/lib/kubelet/kubeconfig file only instead of the whole directory?

Reply:
I'm not kubelet, so I'll defer to @rphillips on that^ question.

Reply:
Yes. We recommend removing the entire /var/lib/kubelet directory.